staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -31,6 +31,77 @@ type DelayingInterface interface {
 	Interface
 	// AddAfter adds an item to the workqueue after the indicated duration has passed
 	AddAfter(item interface{}, duration time.Duration)
+	// AddWithOptions performs an add with additional implementation specific options/features.
+	AddWithOptions(item interface{}, opts AddOptions)
+	// AddAfterWithOptions performs a delayed add with additional implemententation specific options/features.
+	AddAfterWithOptions(item interface{}, duration time.Duration, opts AddAfterOptions)
+	// ForgetDelayed will remove an item from the delayed/waiting queue - it won't cancel active queued items
+	// for them you are still expected to call Done()
+	ForgetDelayed(item interface{}, opts ForgetDelayedOptions)
+}
+
+// AddOptions allows callers to choose from different behaviours when calling AddWithOptions().
+type AddOptions struct {
+	// Waiting specifies how you want the immediate Add to interplay with the existing queue state and
+	// waiting items.
+	Waiting AddVsWaitingBehaviour
+	// Synchronous, when true, instructs AddWithOptions to wait until the item has been proceesed
+	// into the 'active' queue (or dropped) before returning.  Useful for clients/tests that need to
+	// inspect the state of the queue following the add.
+	Synchronous bool
```

Comment:
I agree that it is confusing - because we want to take something that benefits from being highly asynchronous and in both our tests and the tests written for the queue's clients want/expect a synchronous Add because they want to check the queue state immediately after calling it.

Do you have suggestions how to improve this?  I'm running out of ideas.  If we find a way to make both our tests and client tests to always end up with a synchronous action then we are still creating a special code path just for tests, but if we drop the synchronous add behaviour then a whole load of controller tests will immediately fail.

Reply:
Ok I have an idea

Reply:
Please can you take a look at the latest commit.  I've separated out the synchronous setting so that it is only applicable for testing and is private.